### ISSUE TITLE
About the algorithm-problem relationship

### ISSUE THREAD

**Issue by ajnebro** [jMetal/jMetal#55]
I'm opening this issue to start the discussion about the algorithm-problem relationship.

My point of view is the following. An `Algorithm` in jMetal is a metaheuristic, so it is intended to solve a given problem, so I included the relationship "Algorithm solves Problem" in the class diagram.

On the other hand, the interface `Algorithm` does not make any reference to problems:

``` java
public interface Algorithm<R> extends Runnable, Serializable {
  public void run() ;
  public R getResult() ;
}
```

However, if we deep into the class hierarchy, all the `Builder` classes to configure and instantiate algorithms have one mandatory argument: the problem (I discussed about this with Matthieu in another issue), so the relationship Algorithm-Problem is in the code.

I agree that my view is very biased, because the concepts are clear to me. 

We have several ways to proceed:
- Do nothing 
- Consider renaming `Algorithm` as `Metaheuristic´, as in the former versions of jMetal. This vay, the relationship is implicit given the definition of what a metaheuristic is.
- Change the `Algorithm` interface. For example, a `void setProblemToSolve()` could be added, or the `run()` method could be changed to `solve(Problem)`. Here, I like the `run()` method because it allows to run the algorithm in a thread, as Matthie suggested.

This discussion also applies to the relationship "Algorithm uses Operators".

I'm open to suggestions about the most convenient way to proceed.

**Comment by Juanjdurillo** [comment]
I like the second option

**Comment by inaciomdrs** [comment]
Addition of <code>void setProblemToSolve()</code> would be a good idea. Another strategy for this would be using Problem and Operators like push measures (see issue #26).

**Comment by Juanjdurillo** [comment]
The latter option does not seem very logic for me: neither Problem or Operator are measures

**Comment by matthieu-vergne** [comment]
I am currently preparing a big post on that issue, but it could be that it relates also to other concepts in the framework, the same way that here you mentioned the operators. It takes me a while to prepare because I try to support it somehow, and redact it as clearly as possible, because we speak here about central concepts, so it should be precise.

Just a few recommendations:
- to fix biases, go back to the roots: dictionaries and well agreed definitions from communities
- if some people disagree on something, it is probably because they mix several things, including the ones they could agree on: try to split your concepts in order to find which part of it could support the arguments of the others (e.g. I guess inaciomdrs think about a similar registering process to the push measures, thus an observer pattern, not the push measures themselves)

**Comment by inaciomdrs** [comment]
Yes, Matthieu. Organizing better, from the three options initially displayed, I agree with whole second idea (rename <code>Algorithm</code> to <code>Metaheuristic</code>), and a side of third option regarded to provide setter methods for <code>Problem</code>.

**Comment by matthieu-vergne** [comment]
A small comment:

> Consider renaming Algorithm as `Metaheuristic´, as in the former versions of jMetal. This vay, the relationship is implicit given the definition of what a metaheuristic is.

And what is this definition?

From [Wikipedia](http://en.wikipedia.org/wiki/Metaheuristic):
> a metaheuristic is a higher-level procedure or heuristic designed to find, generate, or select a lower-level procedure or heuristic (partial search algorithm) that may provide a sufficiently good solution to an optimization problem, especially with incomplete or imperfect information or limited computation capacity.

Thus, it is a heuristic to provide other heuristics. Thus the output is not problem solutions, but a way to find them.

From Sean Luke, the author of [*Essentials of Metaheuristics*](http://cs.gmu.edu/~sean/book/metaheuristics/), p.9:
> Metaheuristics is a rather unfortunate term often used to describe a major subfield, indeed the primary subfield, of **stochastic optimization**. Stochastic optimization is the general class of algorithms and techniques which employ some degree of randomness to find optimal (or as optimal as possible) solutions to hard problems. Metaheuristics are the most general of these kinds of algorithms, and are applied to a very wide range of problems.
> What kinds of problems? In Jacobellis v. Ohio (1964, regarding obscenity), the United States Supreme Court Justice Potter Stewart famously wrote,
> > I shall not today attempt further to define the kinds of material I understand to be embraced within that shorthand description; and perhaps I could never succeed in intelligibly doing so. **But I know it when I see it**, and the motion picture involved in this case is not that.
> 
> Metaheuristics are applied to *I know it when I see it* problems. They’re algorithms used to find answers to problems when you have very little to help you: you don’t know beforehand what the optimal solution looks like, you don’t know how to go about finding it in a principled way, you have very little heuristic information to go on, and brute-force search is out of the question because the space is too large. *But* if you’re given a candidate solution to your problem, you *can* test it and assess how good it is. That is, you know a good one when you see it.

Thus, a metaheuristic is a stochastic optimization algorithms (i.e. it relies on randomness to provide candidate solutions to a problem) dedicated to deal with problems we don't know much about. Thus, a [travelling salesman problem](http://en.wikipedia.org/wiki/Travelling_salesman_problem) (TSP) is not the kind of problem a metaheuristic is supposed to deal with, because we already know that the solution to a TSP problem is a [minimized cycle over a complete, weighted graph](http://en.wikipedia.org/wiki/Travelling_salesman_problem#As_a_graph_problem).

From [Xin-She Yang](http://www.scholarpedia.org/article/Metaheuristic_Optimization#What_is_Metaheuristic_Optimization):
> Algorithms with stochastic components were often referred to as heuristic in the past, though the recent literature tends to refer to them as metaheuristics. We will follow Glover's convention and call all modern nature-inspired algorithms metaheuristics (Glover 1986, Glover and Kochenberger 2003). Loosely speaking, heuristic means to find or to discover by trial and error. Here meta- means beyond or higher level, and metaheuristics generally perform better than simple heuristics. The word "metaheuristic" was coined by Fred Glover in his seminal paper (Glover 1986), and a metaheuristic can be considered as a "master strategy that guides and modifies other heuristics to produce solutions beyond those that are normally generated in a quest for local optimality" (Glover and Laguna 1997).

Thus, a metaheuristic is a nature-inspired algorithm (I put aside the "modern", which seems to me rather vague) which works via a trial and error process, thus iterative refinements. However, at the same time it takes the position of the *heuristic for heuristics* ("guides and modifies other heuristics") and *heuristic for solutions* ("metaheuristics generally perform better than simple heuristics").

I don't mind for renaming, but the point is not to choose a name and then see that nobody agree on the properties it should cover. If you want a robust implementation, you have to put the properties first, because this is what you will have to implement. Names will only give you class identifiers. So what would you like to deal with? What are the properties which stand out from the old version of jMetal? What are the ones which are generic and the ones which are specific? Which definition do **you** plan to support in this framework? Is this definition concrete enough to allow you to identify the main properties to support? etc.

Exchanging opinions is nice, but if you only rely on informal exchange to draw the main concepts of your framework, you will obtain the same problem than the old versions: a framework which covers many specialized stuff, thus the inability to provide generalized stuff, like experiments. You have to make the requirements clear, otherwise it is like random start in hill-climbing: you don't know where it will go and it can get stuck in some far-to-be-optimal minima. Now, if you plan to make many restarts, probably you will make a close-to-be-optimal solution... but I am not sure of how long people will follow.

There is more or less the same thing in the field related to human intelligence (cf. Cambridge Handbook of Intelligence): more we went, and more perspectives were coming, leading to many debates but no agreements to say how "intelligence" should be defined. For someone like me who wanted to find a good support to know how to define artificial intelligence in a concrete way, it was just a big deception. However, if you look at works on expertise (cf. Cambridge Handbook on Expertise and Expert Performance), which went in parallel but with different assumptions and goals, there is proper convergence and clear definitions, which could maybe be used to identify a proper definition for intelligence (it is one of my projects). In particular because there is a lot of similarities between the two fields.

So the naming is important, but first try to identify the properties. The naming is only there to represent at best these properties. Of course, it is not easy, but this is one of the reasons I am here: to share my experience (and improve it) in Requirements Engineering and the few I have on Ontologies in order to have a clear view of what to consider.

**Comment by ajnebro** [comment]
Matthieu, you are looking at different sources, and I'm affraid that you are getting confused. For example:
* "Thus, a metaheuristic is a nature-inspired algorithm". This is not correct. Scatter search, tabu search, VNS, etc are metaheuristics and they are not nature-inspired.
* "Thus, a travelling salesman problem (TSP) is not the kind of problem a metaheuristic is supposed to deal with". Also incorrect. The TSP is a NP-hard problem, so when the problem size is very large, metaheuristics are techniques able of obtaining quasi-optimal solutions.

I'm not talking as a mathematician, nor I'm giving precise definitions; I'm just a computer science engineer with more then 12 years of experience researching in multi-objetive metaheuristics. And in this research community, a metaheuristic is a non-exact optimization technique aimed at solving an optimization problem. The reference I usually give to my students is this: http://dl.acm.org/citation.cfm?doid=937503.937505

It is about combinatorial problems (as the TSP), but in my opinion is the best introductory paper to metaheuristics in general.

**Comment by matthieu-vergne** [comment]
I am not getting confused. I am well aware that I am not the right one to define what a metaheuristic is, and I intentionally take different sources to show that it is not that clear for someone out from your field (by the way, if you want to identify your biases, you should consider other sources too). If a potential user out of your field wants to know if your framework can provide something useful to him, and see on Wikipedia that metaheuristics are heuristics to select other heuristics, he will probably go away. Names are just identifiers and selling points. My point here is that the discussion should not focus on which name to use, but which properties to provide.

If you have a reference definition, then it is a good thing, but the target then is to see what are the properties that you should infer from that definition in order to identify the main components to put in the highest (generic) interface. Programming is not natural language, at the end you have to implement something, but for that you need to know what, and this is not the name you choose which will tell you that.

Just as an example: your paper mentions:
> Two very important concepts in metaheuristics are intensification and diversification. These are the two forces that largely determine the behavior of a metaheuristic.

If you take it as a reference paper, then I am expecting to see *intensification* and *diversification* as two main concepts in your framework. If not, then I am expecting that you have better to propose. If not, then for generalization purpose I expect to keep speaking about generic algorithms and progressively implements additional features (e.g. builders & factories) to facilitate the implementation of some specific, well known metaheuristics.

Once again, I play the devil advocate on purpose by speaking from the perspective of a new user who has nothing to do with optimization, but who knows that some optimization techniques can help him. If the framework is full of jargon, it cannot be user-friendly.

**Comment by ajnebro** [comment]
> Once again, I play the devil advocate on purpose by speaking from the perspective of a new user who has nothing to do with optimization

I know ;), and you are working a lot, as always, to expose your point of view in a constructive way :)

> If a potential user out of your field wants to know if your framework can provide something useful to him, and see on Wikipedia that metaheuristics are heuristics to select other heuristics, he will probably go away.

Sure, but jMetal, as most of metaheuristic frameworks, is oriented to a especific research area, where everybody speak the same language. So, there is no need to explain e.g., what intensification and diversification mean. The feedback and questions we receive about jMetal are always about technical issues (you can take a look to the LinkedIn user group: https://www.linkedin.com/groups?home=&gid=4441726).

I make sometimes the analogy with Photoshop or Autocad. They are oriented to a specific audience and, when I tried to use the first one, I didn't know the meaning of most of terms and functions.

jMetal is a research tool, so if a user has not a basic knowledge of metaheuristics and multi-objective optimization, then jMetal is not for her/him.

**Comment by matthieu-vergne** [comment]
1 - It is not because you obtain only technical feedback that this is the only thing needed. If people ignore your tool because they don't get what they can do with it, obviously you won't get any feedback from them. Feedback to revise the terminology and relations rarely comes on its own, and it has to be dealt from the main developers of the tool because it has significant impact. From a statistical point of view, if you rely on the feedback to identify the priorities, you consider only the true positives, not the false negatives, although you need both to have a reliable judgement.

2 - When you say:
> jMetal is a research tool, so if a user has not a basic knowledge of metaheuristics and multi-objective optimization, then jMetal is not for her/him.

Where do you place the limit for acceptable "basic knowledge"? For you it is a research tool, but I would like to make it go outside of research, because I would like to use it just because optimization techniques provide efficient solutions. You can think about industrial cases, where all do not have research (or R&D) departments.
 
3 - Notice that I don't mean that you should forbid yourself in using specialized terminology and get stuck with high level generic interfaces: you can freely extend the algorithm interface to have a metaheuristic interface where you will put the interesting stuff you want. If at some point you see that it is too specialized, you still have the possibility to make another interface from the algorithm root and restart. But don't move the root itself to a more specialized level unless you have strong support that it is useless to have it, i.e. you have strong suport that one of your more specialized interface provides the necessary stuff to implement any kind of algorithm that your framework should deal with.

Another point is that our current algorithm interface is not that generic. The getResult() method is rather specific: some algorithms aim at modifying something on the fly, without providing any "result" that they would build. So here we target algorithms which provide concrete outputs, not the ones which just do something, which would be probably the most generic case (thus a simple runnable).

**Comment by ajnebro** [comment]
Well, after thinking about all of this during the week, my proposal is to keep the `Algorithm` interface as is. Two reasons:
* In jMetal there has been always a class `Algorithm`.
* The algorithm is going to optimize a problem, so it must return the best found result.

All the ideas to improve the new versions are welcome, but as it is obvious that each of us have different points of view, we have to take into accout the context of jMetal: we (Juanjo and I) developed it for our own use. Putting it into public domain was a later decision because we thought that our tool could be useful to other people. This is time consuming, and unfortunately we cannot devote to it all the time we would like. 

As all us want to have the version finished in no more than two or three months, we have to be practical: let's fix the most relevant features we think that should be included, and with the gained experience we will have clearer ideas to apply in further versions.

For example, if we consider the measures, the pull measures consisting in getting simple state variable values seems to work pretty well and they are very useful as is. I would fix this feature, and leave for a next release the inclusion of more complex pull measures. About push measures, I have not yet tested it; if they don't work right now, we can pospone them again to a further release.

**Comment by matthieu-vergne** [comment]
Fine with me. I keep what I have prepared to feed this discussion (I will finish it when I have time) and we can go ahead at our pace, keeping it as a parallel discussion.

**Comment by ajnebro** [comment]
After working in adding support for experimental studies I have found that all the algorithms require a method to get the problem to solve.

This an old discussion, but the point is that I'm thinking in changing the current `Algorithm` class:

```java
public interface Algorithm<Result> extends Runnable, Serializable, DescribedEntity {
  void run() ;
  Result getResult() ;
}
```

by this one:
```java
public interface Algorithm<S extends Solution<?>, Result> extends Runnable, Serializable, DescribedEntity {
  void run() ;
  Result getResult() ;
  Problem<S> getProblem() ;
}
```

**Comment by ajnebro** [comment]
By the way, the `Algorithm` interface now extends `DescribedEntity`. This latter interface is becoming very useful and I will probably add it to other core classes (`Problem` and `Operator` at least).

**Comment by matthieu-vergne** [comment]
Once again, you go back to the idea that an algorithm is made to solve a given problem, which is obviously not the case when you think about algorithms like NSGA2, which can be applied to many different problems. What makes it problem-specific is how you will use/interprete the solutions returned, and how you will set up the parameters of the algorithm. But the algorithm itself is not designed to deal with a specific problem instance. So introducing this method here seems to me a huge mistake, because you introduce unmotivated dependencies, and moreover in a core class, which means that all the framework will then rely on it and it will be hard to fix on the long term.

Could you please describe the contexts which requires it? So I can give you my view on where this method could make sense.

**Comment by matthieu-vergne** [comment]
OK, so let's come back to this issue. Time is our friend, so here is a new proposal:
```
public interface Algorithm<Result> extends Serializable, DescribedEntity {
  <S extends Solution<?>> void run(Problem<S> problem) ;
  Result getResult() ;
}
```
Here, the point is not that the algorithm is designed for a specific problem (so it is not the algorithm which provides the problem) but it has to take care of problem specificities to be run, in particular because it must solve the problem. This is why it should *receive* the problem to solve.

Here I continue the discussion started in #135 about that.

> The fast answer would be "alternative 2), of course", but this option wouldn't allow to get information of the problem in the algorithm class constructor.

But this is the whole point: if you need to know what problem to solve before to set up anything for an algorithm, then your algorithm is problem specific. To make it problem independent, you have to (i) reduce what you actually take from the problem, and (ii) delay it as much as possible.

I often see in your code `problem.xxx` for obtaining data that you can get in another way, so I am pretty sure that you can remove a lot of these call to the problem. Moreover, considering that the algorithm is not supposed to do anything before to be run in general, it seems to me that delaying the use of the problem to the `run()` method is clearly feasible.

Just an example with `AbstractMOEAD`. If you check the constructor, here is all what you need it for:
```
indArray = new Solution[problem.getNumberOfObjectives()];
idealPoint = new double[problem.getNumberOfObjectives()];
nadirPoint = new double[problem.getNumberOfObjectives()];
lambda = new double[populationSize][problem.getNumberOfObjectives()];
```
These can be done at the beginning of `run()`. As a matter of fact, you can directly integrate these instanciations in the corresponding init functions:
- `initializeUniformWeight()` (fill `lambda`)
- `initializeIdealPoint()` (fill `idealPoint`)
- `initializeNadirPoint()` (fill `nadirPoint`)

Because actually, instanciating them is nothing more than the first init step before to fill them with default values. So as soon as you move them in their init functions, only `indArray` remain. So how to extract it from the constructor?

Well, same thing: just integrate it to the `run()` method. But here is the problem: you do not implement this method here because you let it for the specific implementations to do it. The trick here is that you require your implementations to use your init functions by themselves. So when you say:
> NSGA-III and most of algorithms implements a template that does not have any method for initialization. We should then include something like an init() method in all the templates ... I'm not sure to like that idea ...

The point is that you already do it here. You don't like it? I don't either, so here is a solution. If you define methods, the only way you can be sure that they are properly called is to control these calls by yourself. Here you give the opportunity for the implementation to make a mistake by not initializing them properly. In your abstract class, you can define:
```
@Override
public void run() {
	initializeUniformWeight();
	initializeNeighborhood();
	initializeIdealPoint();

	executeMOEAD();
}

protected abstract void executeMOEAD();
```
So the implementation has to implement `executeMOEAD()` instead of `run()`, and even if you make init methods to organize your code, just make them private: your implementations don't have to care about them. This kind of structure not only ensures that you call your init methods at the right time, but also that you provide all the algorithm-specific stuff to the implementation through the arguments of your new method:
```
executeMOEAD(idealPoint, lambda, ...);
```
so that you only need local variables instead of global ones, which gives you a complete guarantee that, if you re-run the algorithm on another problem, then you are sure not to have side effects due to a previous run.

Another thing to mention: there is not only these functions which need the problem, but these ones too:
- `updateNadirPoint()` (replaceable by `nadirPoint.length`)
- `updateIdealPoint()` (replaceable by `lambda[0].length`)
- `fitnessFunction()` (replaceable by `idealPoint.length`, `lambda.length`)

But if I am right, because they all use one of these problem-based variables instanciated above, none of them actually need the problem itself. Actually, you can replace all the problem calls of these functions:
- `updateNadirPoint()` replaceable by `nadirPoint.length`
- `updateIdealPoint()` replaceable by `lambda[0].length`
- `fitnessFunction()` replaceable by `idealPoint.length` and `lambda.length`

So here is another evidence that you are just over-constraining yourself in using your problem. The problem has few uses:
- initialize your variables
- generate your solutions
- evaluate your solutions

And all of them can be done (and are done anyway) in the `run()` method. The only things you have out of it are your instanciations, which can be moved to the run method and, in this way, offers you to rely on local variables for less side effects.

> We have to consider also that we usually use builder classes to set the algorithm parameters, and the first parameter of the builder constructors is the problem. 

If you tell me that your builder needs more parameters than what your algorithm needs, then I can tell you that you have something different than a mere builder. If we move the problem to the run method, there is no reason for your builders to keep it.

So as a summary: move many -if not all- of your constructor stuff to the run method, use local variables, use `run(problem)` instead of `run()` and you have it. The algorithm *definition* is independent of the problem, but you have to *run* it on a given problem to make sense.

What do you think?

**Comment by ajnebro** [comment]
>> The fast answer would be "alternative 2), of course", but this option wouldn't allow to get information of the problem in the algorithm class constructor.

> But this is the whole point: if you need to know what problem to solve before to set up anything for an algorithm, then your algorithm is problem specific.

I agree. In fact, I'm not saying that getting problem information in the algorithm constructor is a desirable feature; I'm merely stating that currently some algorithm implementations do that.

**Comment by ajnebro** [comment]
> ... if you re-run the algorithm on another problem, then you are sure not to have side effects due to a previous run.

You are starting to convince me ;)

**Comment by ajnebro** [comment]
Certainly, replacing state variables by local variables helps in the sense of making the algorithms thread-safe, which is good.

Now we should think in the consequences of adopting this scheme in the current framework. For instance, the `Algorithm` interface does not extends `Runnable`, and we are assuming that in all the runner classes.

**Comment by matthieu-vergne** [comment]
But I bet that this is only a technical issue. Where do you use this property?

**Comment by matthieu-vergne** [comment]
Oops, wait, I will check the runner classes. (so same claim, but I will check myself)

**Comment by matthieu-vergne** [comment]
Lol. If I remove `Runnable`, there is only 2 issues:

The first one is with [line 47 in `AlgorithmRunner`](https://github.com/jMetal/jMetal/blob/master/jmetal-core/src/main/java/org/uma/jmetal/util/AlgorithmRunner.java#L47).

**Solution 1:** replace `new Thread(algorithm)` by using a proper `Runnable` to enclose the algorithm run, the issue being to add the problem, what can be done simply by requesting it as an argument of the `execute()` method. The 63 errors due to such a change are trivial: in every cases, you are in a situation where you have both the algorithm instance and the problem instance, so there is no difficulty, just provide the problem you have. Actually, you are always in a case where you instantiate the algorithm just few lines above, so you need the problem there.

**Solution 2:** rather than providing an algorithm, you should provide a pair (algorithm, problem), which is fundamentally the same, just give it another looking. The point is that it highlight the fact that this is not a simple "algorithm runner" but a "problem solver" (run an algorithm on a given problem to solve it). But out of this conceptual shift, technically speaking it is strictly equivalent to the solution 1.

The second issue is with [line 120 in `NSGAIIMeasuresRunner`](https://github.com/jMetal/jMetal/blob/master/jmetal-exec/src/main/java/org/uma/jmetal/runner/multiobjective/NSGAIIMeasuresRunner.java#L120). Here, nothing tricky : the problem is just above, so solution 1 directly: use a proper `Runnable` and just provide the problem already there.

**Comment by matthieu-vergne** [comment]
As a matter of fact, instead of `run(problem)` you can rename it `solve(problem)` to be compliant with your model.

**Comment by matthieu-vergne** [comment]
And before you do any big change, consider that we can go progressively without removing anything to reduce the impact on pzople using jmetal.

**Comment by ajnebro** [comment]
> As a matter of fact, instead of `run(problem)` you can rename it `solve(problem)` to be compliant with your model.

I agree.

**Comment by ajnebro** [comment]
> So far I do not remember any case (besides the probability for mutation) when a constructor needed problem info.

I'm thinking in this sentence of Juanjo in issue #135, which links with one of you in the same issue:
> The second allows you to set up everything, and at the last moment to decide on which problem you apply it, which means that nothing forbids you to apply the same algorithm instance to another problem just after.

Most of metaheuristics use a mutation operator, which usually is applied with a probability `pm = 1.0/problem.getNumberOfVariables()`. This means that we could lose one of the advantages of the new scheme, because a created algorithm could not be applied to solve different problems unless they have the same number of variables.

**Comment by matthieu-vergne** [comment]
My opinion is that you should stop mixing everything and think about what pertains to the algorithm, what pertains to the problem, etc. Whether you have the will to decrease the dependencies and you design your classes consequently, whether you assume that you have some flow (first this, then this, etc.) and you target a significantly different design.

No you don't loose anything: if you have to set up a parameter based on an information that you do not have yet, then don't ask it yet. The operator is the operator, the probability for it to be applied does not depend on the operator itself, it depends on how you use it. It has nothing to do in the definition of the operator. So extract that from the operator to give it back to the algorithm and you will see it fits.

**Comment by ajnebro** [comment]
Hmmm, I like the idea although I don't see right now how to do it.

Anyway, adopting this new scheme would involve changing at least all the algorithms and operators. These are significant changes and I don't think that they should be applied to the current development version 5.1, which is supposed to be a minor revision. Furthermore, I have three projects using jMetal5.1 as a dependency and then I should modify them. 

I'm thinking in freezing the features of version 5.1 after we decide how to manage the issue of the experiments. Once we release jMetal 5.1 then we would be free to start a major revision.

**Comment by matthieu-vergne** [comment]
I think that there is a lot to do before to have a clean experiment architecture. So I recommend that you go with your current experiments and let the major revision for later. Because so far you have too much dependencies to properly cut it, and this should be solved before to have experiments building on that. I can try to analyse some specific issues that you would like to be solved for the minor revision, but you will have to make dedicated issues so we can keep the general discussions at the current state.

**Comment by matthieu-vergne** [comment]
I think that we can go progressively, at least for some things, by only adding stuff and keeping the old stuff deprecated (not removing it) but we still need to make changes at high levels, so we need then to apply it everywhere.

**Comment by matthieu-vergne** [comment]
Is the 5.1 release done? Do you think we can work on a major revision now? I am currently working on some operator-based templates, especially for issue #110, and I am -of course- running through the dependence to the problem, which for now is nothing more than a syntactical optimization (you factor a common code for some child classes in a parent class which actually does not need it, and thus imposes it on child classes which do not need it).

If we can go for replacing `run()` by `solve(problem)`, I can start on it.

**Comment by matthieu-vergne** [comment]
I have started to work on this change locally, and the point is that your templates are completely oblivious to this problem, which means that whether your problem is, practically, useless at this level, whether your should revise your templates to consider it. Otherwise it just add to my own position, which is that an algorithm does not depend on a problem. Actually, so far the problem variables that you included in high level classes, like in AbstractEvolutionaryAlgorithm, are mere syntactical optimizations (you factor a code common to some child classes and impose it on all the hierarchy). I was actually ready to have the same thing that my pull request #159 but for problems.

So if you want to go ahead with integrating the problem into the algorithm, which would make sense if we take your model as a reference, it means that some other modifications should be considered too to integrate it properly.

**Comment by matthieu-vergne** [comment]
For instance, in `AbstractScatterSearch`, the run method becomes the following:
```java
  @Override public void solve(Problem<S> problem) {
    initializationPhase() ;
    referenceSetUpdate();
    while (!isStoppingConditionReached()) {
      List<List<S>> subset = subsetGeneration();
      List<S> combinedSolutions = solutionCombination(subset) ;
      if (restartConditionIsFulfilled(combinedSolutions)) {
        restart();
        referenceSetUpdate();
      } else {
        for (S solution : combinedSolutions) {
          S improvedSolution = improvement(solution);
          referenceSetUpdate(improvedSolution);
        }
      }
    }
  }
```
The problem being unused, any child class which has to implement the methods required by this template do not receive the problem (and making a setter/getter for it just break the purpose of focusing the problem on the run). The methods should be revised to use the problem provided, otherwise it makes no sense.

**Comment by matthieu-vergne** [comment]
It seems that `CovarianceMatrixAdaptationEvolutionStrategy` is particularly pathological case for the adaptation for two reasons.

One is that it uses a builder additionally to the problem. Actually the problem is obtained from the builder, but basically it is given to it before to be retrieved by the algorithm itself, without being used at all by the builder, so you practically have the problem and, independently to it, a builder. The second reason is that it makes a lot of initialization bazed on the problem and the builder and need to update them at each iteration. What makes it a problem is that it extends `AbstractEvolutionStrategy`, so if the initialization could be done easily by overriding the solve method to make the initialization just before to run the solve method of the parent, the regular update however does not have any method to rely on. Currently, it is done during the `updateProgress()` method, which is confusing (because it aims at dealing with the progress, not something else) but is the most suited place to trick the algorithm... which obviously mean that you are playing with the fire here, because you use a class in a way it is not intended to be used.

I would recommend first to extract this builder from there and revise it, so the `CovarianceMatrixAdaptationEvolutionStrategy` does not depend on this builder. It seems the builder only stores data and build the algorithm by providing the builder itself in argument while it could just provide the data directly, like other builders, so it does not make sense to me to provide the builder. Revising it to be similar with the other builders (including by extending the `AlgorithmBuilder` interface) seems to me a good idea. This would remove a weird dependency between the two classes, and would allow to clearly see what should be provided in the constructor and what can be preserved for the running. This is what I tried to do in my pull #160.

Second, I would recommend to think deeply about what this algorithm involves. In particular, is is true that the data stored in the builder (excepted the problem) must be required from the constructor, or is it true that we could wait until the call of the run method? So far, the builder provide default values, and its only use is in `CovarianceMatrixAdaptationEvolutionStrategyRunner` which just do nothing else than instantiate it and build the algorithm, so exploit the default value. Such default values can be simply put in the algorithm itself rather than a builder, and the setters/getters of the builder can be integrated as well, which seems to me to simplify a lot.

**Comment by matthieu-vergne** [comment]
Considering that a significant part of the discussion was about parameters of the algorithm which need to be set at various times depending on the context, would you be interested to have setter/getters for each parameter? The idea is to be able to set them whenever we need/want before the execution of the `Algorithm`. If the `Algorithm` is coded for it, it is also possible to change the parameters during the run, like required in #185.

**Comment by matthieu-vergne** [comment]
Basically, I am speaking about what I made in #72. It was merged and then lost during the cleaning for the frist 5.0 release. Maybe it would be time to get it back.

**Comment by ajnebro** [comment]
I agree that it is worth revisiting this issue.

**Comment by matthieu-vergne** [comment]
I am currently working on the experiment architecture. It is well advanced, but not finished yet (I don't have a full experiment ready yet). I will present it in #135 when it will be complete, but so far it does not depend on the dependency between algorithms and problems (they can be completely independent or completely coupled, it works anyway) because it requires only the algorithm (the problem can be used in any way, the experiment architecture is oblivious to it). This way, it offers a lot of flexibility, so we can use it already with the current jMetal architecture, and even if we revise the dependency between algorithms and problems it should not impact the architecture of experiments.
